community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala

```diff
@@ -74,9 +74,19 @@ abstract class CompiledExecutionResult extends InternalExecutionResult {
 
   override def queryStatistics(): InternalQueryStatistics = InternalQueryStatistics()
 
+  private var successful = false
+  protected def success(): Unit = {
+    successful = true
+  }
+
   override def close(): Unit = {
-    ensureIterator()
-    innerIterator.close()
+    if (innerIterator != null) {
+      innerIterator.close( )
+    }
+    if (statement != null) {
```

Comment:
using an option for the `statement` and `innerIterator`? so the `if`s can be replaced by `foreach`s